### PR TITLE
feat: makes job timeout resolution configurable

### LINKED ISSUE #5073: Increase job timeout resolution / reject requests below resolution threshold
At the moment, the broker job timeout resolution is 30 seconds. In the client library I can set the `timeout` for an ActivateJobs request that is less than 30 seconds.

However, when I run my program, I see retries happening every 30 seconds. Without understanding the internal mechanics of the broker, or reading the documentation to discover how it works, the behaviour is unexpected and inexplicable.

**Describe the solution you'd like**
Three potential approaches: 

1. Increase the broker timeout resolution to 1s. This may have a performance impact on the operation of the broker. This requires no change to the client libraries, and more closely aligns with an intuitive, unconscious expectation of users.
2. If not this, then code the client libraries so that the timeout has to be specified using a custom class that explicitly takes intervals of 30s - eg: `Job.Timeout30SecondBlock(1)`. This makes the client a little more complex, but the behaviour is semantically explicit. It makes getting started a little more complicated, but pushes the validation to the IDE and explicit and discoverable.
3. Throw an error when an ActivateJobs request is received with a timeout that is either less than 30 seconds, or not a multiple of 30 seconds. The error message would say: `Job timeout must be a multiple of 30 seconds`. This pushes the validation to run-time, but is explicit and discoverable. This behaviour means less impact on client libraries, and is easier to change in the future without requiring changes to the client libraries.

### PR DESCRIPTION
**Pull request by mustafadagher**
## Description

Added configs for JobTimeoutTrigger polling interval, and a batch limit to manage any possible performance degradation

## Related issues

closes #5073 

## Definition of Done

_Not all items need to be done depending on the issue and the pull request._

Code changes:
* [x] The changes are backwards compatibility with previous versions
* [x] If it fixes a bug then PRs are created to [backport](https://github.com/camunda/zeebe/compare/stable/0.24...main?expand=1&template=backport_template.md&title=[Backport%200.24]) the fix to the last two minor versions. You can trigger a backport by assigning labels (e.g. `backport stable/1.3`) to the PR, in case that fails you need to create backports manually.

Testing:
* [x] There are unit/integration tests that verify all acceptance criterias of the issue
* [x] New tests are written to ensure backwards compatibility with further versions
* [x] The behavior is tested manually
* [ ] The change has been verified by a QA run
* [ ] The impact of the changes is verified by a benchmark

Documentation:
* [ ] The documentation is updated (e.g. BPMN reference, configuration, examples, get-started guides, etc.)
* [ ] If the PR changes how BPMN processes are validated (e.g. support new BPMN element) then the Camunda modeling team should be informed to adjust the BPMN linting.

Other teams:
If the change impacts another team an issue has been created for this team, explaining what they need to do to support this change.
- [ ] [Operate](https://github.com/camunda/operate/issues)
- [ ] [Tasklist](https://github.com/camunda/tasklist/issues)
- [ ] [Web Modeler](https://github.com/camunda/web-modeler/issues)
- [ ] [Desktop Modeler](https://github.com/camunda/camunda-modeler/issues)
- [ ] [Optimize](https://github.com/camunda/camunda-optimize/issues)

Please refer to our [review guidelines](https://github.com/camunda/zeebe/wiki/Pull-Requests-and-Code-Reviews#code-review-guidelines).

### CODE REVIEW

File: zeebe/engine/src/main/java/io/camunda/zeebe/engine/EngineConfiguration.java
Comment (mustafadagher) [inline zeebe/engine/src/main/java/io/camunda/zeebe/engine/EngineConfiguration.java:17]: I know in the ticket there were comments about using 1s resolution by default, but I was not so sure how I feel about changing the default behavior suddenly.
I'm open for any advice to change it though

File: dist/src/main/config/broker.standalone.yaml.template
Comment (mustafadagher) [inline dist/src/main/config/broker.standalone.yaml.template:1156]: This is probably missing some more mature documentation, I tried my best according to my knowledge, but comparing to the one above it (from `messages`), it could use some help!

File: zeebe/engine/src/main/java/io/camunda/zeebe/engine/state/instance/DbJobState.java
Comment (megglos) [inline zeebe/engine/src/main/java/io/camunda/zeebe/engine/state/instance/DbJobState.java:452]: 🔧  what's the reason for extraction into a dedicated method? I also think the simple name `visit` is not really a good fit in this case as this is a specific visit logic only relevant for the only calling method `forEachTimedOutEntry` at least `visitTimedOutJob` would be a better fit if you really need to extract it

File: zeebe/engine/src/main/java/io/camunda/zeebe/engine/EngineConfiguration.java
Comment (megglos) [reply zeebe/engine/src/main/java/io/camunda/zeebe/engine/EngineConfiguration.java:17]: I think for `main` that's totally ok, for backporting we may keep the old default to not introduce a behavior change

**Review by megglos — CHANGES_REQUESTED** [review]
Great job! Mostly naming comments :D and I think we should actually change the default to provide a better experience by default with the upcoming 8.5 release.

Other than that I wonder if it's really worth having a configurable batch size, that was also not considered in the original scope of the issue but I assume you introduced it for alignment with the similar message expiry config. I would like to get the perspective of @oleschoenburg  and @korthout  there.

One more thing, given the interval is configurable now, do you see an easy way to add a test case that verifies set config interval is indeed effective? That would be great to ensure all is and stays wired up as expected.

Would also await a review by at least one other engineer, as my knowledge is limited in this area.

File: dist/src/main/config/broker.standalone.yaml.template
Comment (oleschoenburg) [reply dist/src/main/config/broker.standalone.yaml.template:1159]: It adds a decent amount of complexity but since it's already implemented now and it's nice to have for fairness, I'd suggest to keep it now.

:thought_balloon: In general I think it doesn't make much sense to keep these configuration flags as experimental; I don't see how and when we'd decide to promote them to stable. This isn't something we need to change in this PR though.

File: zeebe/engine/src/main/java/io/camunda/zeebe/engine/state/immutable/JobState.java
Comment (oleschoenburg) [inline zeebe/engine/src/main/java/io/camunda/zeebe/engine/state/immutable/JobState.java:55]: :+1: Nice that you added a type for that!

:wrench: Since `JobState` tracks more than just deadlines, I'd suggest to rename it to `DeadlineIndex` or something similar.

File: zeebe/engine/src/main/java/io/camunda/zeebe/engine/processing/job/JobTimeoutTrigger.java
Comment (oleschoenburg) [reply zeebe/engine/src/main/java/io/camunda/zeebe/engine/processing/job/JobTimeoutTrigger.java:92]: Or `deadline` or `untilTimestamp` or `expiredBefore`...

**Review by lenaschoenburg — APPROVED** [review]
Really nice job @mustafadagher :+1: 

:upside_down_face: Feel free to make more use of `var` or `final var` when declaring variables.

File: dist/src/main/config/broker.standalone.yaml.template
Comment (megglos) [reply dist/src/main/config/broker.standalone.yaml.template:1157]: gave a good laugh yesterday as well and I sympathize with this type of type as my fingers also sometimes mix up first letters of sequential words :D

Comment (mustafadagher) [reply dist/src/main/config/broker.standalone.yaml.template:1157]: I laughed really loud seeing the comment, almost made my morning haha 😄

File: zeebe/engine/src/main/java/io/camunda/zeebe/engine/state/instance/DbJobState.java
Comment (mustafadagher) [reply zeebe/engine/src/main/java/io/camunda/zeebe/engine/state/instance/DbJobState.java:346]: That's a good one. It also reminds me that I wanted to use the `batchLimitExceeded` somewhere outside the loop, but I can't recall where now 😄

File: zeebe/engine/src/main/java/io/camunda/zeebe/engine/state/instance/DbJobState.java
Comment (mustafadagher) [reply zeebe/engine/src/main/java/io/camunda/zeebe/engine/state/instance/DbJobState.java:452]: I just thought that the inline method inside `deadlinesColumnFamily.whileTrue` got really longer and unreadable. I will rename the method

File: zeebe/engine/src/main/java/io/camunda/zeebe/engine/processing/job/JobTimeoutTrigger.java
Comment (korthout) [inline zeebe/engine/src/main/java/io/camunda/zeebe/engine/processing/job/JobTimeoutTrigger.java:106]: 🔧 We could add some trace level logs every time it runs and debug level when it yields early. 

@megglos @mustafadagher What do you think?

**Review by korthout — CHANGES_REQUESTED** [review]
Nice work @mustafadagher 👏

👍 Very impressive that you were able to build a complete solution already!

I have a few requests and suggestions. Please have a look.

❌ To resolve the bug, we should change the default both on `main` and backports. Let's go for `1s` by default. Please backport this to all supported versions `8.4`, `8.3`, `8.2`, and `8.1`.

🔧 Let's refer to the concept as the job timeout checker, both in code and documentation. The trigger term is legacy and is no longer used. These classes are typically referred to as checkers by the team. We should align our code and concepts to avoid confusion.

💭 The number of changes in this single commit pull request makes it challenging to review. I suggest splitting the PR into smaller commits or separate PRs to [help the reviewer](https://github.com/camunda/zeebe/wiki/Pull-Requests-and-Code-Reviews).

File: zeebe/engine/src/main/java/io/camunda/zeebe/engine/EngineConfiguration.java
Comment (korthout) [reply zeebe/engine/src/main/java/io/camunda/zeebe/engine/EngineConfiguration.java:17]: IMO, we should resolve the bug on all supported versions. Let's use 1s on all versions by default.

File: dist/src/main/config/broker.standalone.yaml.template
Comment (korthout) [reply dist/src/main/config/broker.standalone.yaml.template:1159]: @oleschoenburg @megglos I'd like to extract this functionality and apply it to all checkers to achieve fairness and flow within the periodically executed scheduled tasks.

I imagine having a single config option applied to all checkers equally rather than individual controls. Together with metrics to support tweaking this config for different usages.

While this solution uses a number of resulting records appended to the log in a batch, an alternative would be based on execution time. I'm not sure yet which is better, or whether both might be needed at some point.

Anyways, I instructed @mustafadagher to go for this solution and take the same approach as the Message TTL Checker as it resolves the bug in a safe way and controllable way.

File: zeebe/engine/src/test/java/io/camunda/zeebe/engine/state/instance/JobStateTest.java
Comment (korthout) [reply zeebe/engine/src/test/java/io/camunda/zeebe/engine/state/instance/JobStateTest.java:561]: Already mentioned by Ole above

File: zeebe/engine/src/main/java/io/camunda/zeebe/engine/processing/job/JobTimeoutTrigger.java
Comment (oleschoenburg) [reply zeebe/engine/src/main/java/io/camunda/zeebe/engine/processing/job/JobTimeoutTrigger.java:106]: I'm fine with the trace logs, debug is too noisy IMO.

File: zeebe/engine/src/main/java/io/camunda/zeebe/engine/state/immutable/JobState.java
Comment (megglos) [reply zeebe/engine/src/main/java/io/camunda/zeebe/engine/state/immutable/JobState.java:26]: nice catch! having the logic on the same component level would be worth doing now.

Comment (megglos) [reply zeebe/engine/src/main/java/io/camunda/zeebe/engine/state/immutable/JobState.java:26]: I thinks it's doing while we are here instead of duplicating the logic further

File: zeebe/engine/src/test/java/io/camunda/zeebe/engine/state/instance/JobStateTest.java
Comment (mustafadagher) [reply zeebe/engine/src/test/java/io/camunda/zeebe/engine/state/instance/JobStateTest.java:597]: good catch! Will cover the first case.
The second was already covered by all the existing test having `Integer.MAX_VALUE` passed as a batch limit

File: zeebe/engine/src/main/java/io/camunda/zeebe/engine/state/immutable/JobState.java
Comment (mustafadagher) [reply zeebe/engine/src/main/java/io/camunda/zeebe/engine/state/immutable/JobState.java:26]: That's what I also think. I will move the batch limit in my next commit. but extracting the logic could use a new issue.

File: zeebe/engine/src/main/java/io/camunda/zeebe/engine/processing/job/JobTimeoutChecker.java
Comment (mustafadagher) [inline zeebe/engine/src/main/java/io/camunda/zeebe/engine/processing/job/JobTimeoutChecker.java:93]: @korthout I've moved the batchLimit comparison here and passed a `BooleanSupplier` instead. I have a feeling that there could have been a better solution, but this does the job. WDYT?

CC: @megglos

Comment (korthout) [reply zeebe/engine/src/main/java/io/camunda/zeebe/engine/processing/job/JobTimeoutChecker.java:93]: Currently, this does not allow extracting the yielding/restarting-at-index behavior for reuse. The `exceedsBatchLimit` also doesn't fit the abstraction of the JobState, which only provides data access (i.e. it does not know what a batch or a batch limit is).

I suggest only passing the `timestamp`, `index` and `visitor` to the `forEachTimedOutEntry`. This is the bare minimum needed for data access.

The batch limit can be handled inside the visitor, like the message ttl checker. By aligning the implementations already, we can more easily extract the logic for reuse in all checkers.

Comment (mustafadagher) [reply zeebe/engine/src/main/java/io/camunda/zeebe/engine/processing/job/JobTimeoutChecker.java:93]: I see what you mean.
The main difference between this and the messageTtlChecker that made it difficult to align both implementation is that this method here `forEachTimedOutEntry` receives a. `callback` method that already returns a `boolean` and that is used in some places. the other one was initially a void and it was easier to manipulate it to fix this.

I'll look into it more by all means

**Comment by mustafadagher** [comment]
I cleaned the git history (or at least, I tried) for fun (and also after a really insightful conversation with @korthout) :bowtie: :sparkles:
FYI: @megglos @oleschoenburg

File: zeebe/engine/src/main/java/io/camunda/zeebe/engine/processing/job/JobTimeoutTrigger.java
Comment (mustafadagher) [reply zeebe/engine/src/main/java/io/camunda/zeebe/engine/processing/job/JobTimeoutTrigger.java]: :thought_balloon: just an observation, none of the other checkers in the engine log anything. was that some kind of a design decision?

Comment (oleschoenburg) [reply zeebe/engine/src/main/java/io/camunda/zeebe/engine/processing/job/JobTimeoutTrigger.java]: I don't think so, just something that we haven't bothered to do before. You can just copy how we do logging in other classes, nothing special for scheduled tasks necessary.

File: zeebe/engine/src/main/java/io/camunda/zeebe/engine/processing/job/JobTimeoutChecker.java
Comment (mustafadagher) [reply zeebe/engine/src/main/java/io/camunda/zeebe/engine/processing/job/JobTimeoutChecker.java:92]: 💭 instead or in addition?

Comment (mustafadagher) [reply zeebe/engine/src/main/java/io/camunda/zeebe/engine/processing/job/JobTimeoutChecker.java:92]: and do you think something like this be sufficient? or what would you like to see?
```java
LOG.trace("{} jobs has been marked to timeout", counter.get());
```

Comment (megglos) [reply zeebe/engine/src/main/java/io/camunda/zeebe/engine/processing/job/JobTimeoutChecker.java:92]: I would be in favor of both, would also allow us to understand how long it took from the log timestamps

File: zeebe/engine/src/main/java/io/camunda/zeebe/engine/processing/job/DeactivateTimeOutJobs.java
Comment (korthout) [inline zeebe/engine/src/main/java/io/camunda/zeebe/engine/processing/job/DeactivateTimeOutJobs.java:24]: 🔧 I suggest we rename this to `JobTimeoutChecker` as it fully takes the role of a checker, including rescheduling itself. Let's then also rename the current `JobTimeoutChecker` to something else, as it only manages the lifecycle of the checker now.

**Review by korthout — CHANGES_REQUESTED** [review]
Well done @mustafadagher 👏 Almost there!

👍 I really like how you split up the commits. It was a pleasure to review!

👍 I like that you set up the `DeactivateTimeOutJobsTest` for easy testing!

💭 I'm not a big fan of testing with mockito as it typically verifies implementation rather than behavior, but I think it's reasonable for this class. The setup of each test case made sense to me while reviewing. Also, the verifications are done on the interface between the stream processor and the checker, which is less likely to change than the rest of the implementation.

❌ I noticed a singular naming issue we should change as it leaks into the interface documentation. Please adjust it.

🔧 I also saw some possible minor improvements. Please have a look.

❓ I wonder whether we should change the default batch limit from MAX_INT to something more useful. But we can also do this separately. Curious to your thoughts about this.

File: zeebe/engine/src/main/java/io/camunda/zeebe/engine/processing/job/DeactivateTimeOutJobs.java
Comment (megglos) [inline zeebe/engine/src/main/java/io/camunda/zeebe/engine/processing/job/DeactivateTimeOutJobs.java:55]:
```suggestion
    LOG.trace("Job timeout checker running...");
```

File: zeebe/engine/src/main/java/io/camunda/zeebe/engine/processing/job/DeactivateTimeOutJobs.java
Comment (megglos) [inline zeebe/engine/src/main/java/io/camunda/zeebe/engine/processing/job/DeactivateTimeOutJobs.java:76]:
```suggestion
          "Job timeout checker yielded early. will reschedule immediately from where it left of.");
```

File: zeebe/engine/src/main/java/io/camunda/zeebe/engine/processing/job/DeactivateTimeOutJobs.java
Comment (megglos) [inline zeebe/engine/src/main/java/io/camunda/zeebe/engine/processing/job/DeactivateTimeOutJobs.java:85]:
```suggestion
    LOG.trace("{} jobs have been marked to timeout.", counter.get());
```

File: zeebe/engine/src/main/java/io/camunda/zeebe/engine/processing/job/JobTimeoutChecker.java
Comment (megglos) [inline zeebe/engine/src/main/java/io/camunda/zeebe/engine/processing/job/JobTimeoutChecker.java:17]: 🔧 As this class now cares about scheduling the checker triggered by the stream processor lifecycle, wdyt about naming it `JobTimeoutCheckerScheduler`? 🤔

**Review by megglos — APPROVED** [review]
Good job! 🚀 
Nothing to add beyond what @korthout raised, one naming suggesting for the class that takes care of scheduling the checker triggered by the stream processor lifecycle.

Pre-approved!

File: zeebe/engine/src/main/java/io/camunda/zeebe/engine/processing/job/DeactivateTimeOutJobs.java
Comment (mustafadagher) [reply zeebe/engine/src/main/java/io/camunda/zeebe/engine/processing/job/DeactivateTimeOutJobs.java:24]: I also had the same feeling but I couldn't know what to rename the actual class to, so thought of keeping the current names as is.
Now I'm thinking the other one can be renamed to `JobTimeoutListner` or `JobTimeoutObserver `, WDYT?

File: zeebe/engine/src/main/java/io/camunda/zeebe/engine/processing/job/JobTimeoutChecker.java
Comment (mustafadagher) [reply zeebe/engine/src/main/java/io/camunda/zeebe/engine/processing/job/JobTimeoutChecker.java:17]: I was suggesting `JobTimeoutListner` or `JobTimeoutObserver` on a comment above, but `JobTimeoutCheckerScheduler` seems reasonable too!

File: zeebe/engine/src/main/java/io/camunda/zeebe/engine/state/instance/DbJobState.java
Comment (korthout) [reply zeebe/engine/src/main/java/io/camunda/zeebe/engine/state/instance/DbJobState.java:355]: That's not correct @mustafadagher. Consider the interface of this method here. Is the visitor called for this index yes/no? It is, but the visitor returned false, indicating that it stops here. The caller then receives the last visited index back so it can use it to start there again next time. This is the last visited index and the one to start with next time.

But it's not the next index. Because that would mean that the visitor did not yet see this entry. However, it did already see it and it responded false to it.

Comment (mustafadagher) [reply zeebe/engine/src/main/java/io/camunda/zeebe/engine/state/instance/DbJobState.java:355]: okay, I'm kinda convinced haha. will change it :)

**Comment by korthout** [comment]
>I think the best way to approach this is to monitor both the performance with indefinite limit and the errors if any occurs for some period (say 1-3 months), then if we see any alarming degradation, we would have more data to be able to decide what can be a reasonable limit. we could also experiment there; e.g. change it down bit by bit and see what works best.

I agree. This would be easier if we have metrics on run durations of scheduled tasks. @megglos is that something we could ask from ZDP?

**Comment by megglos** [comment]
> > I think the best way to approach this is to monitor both the performance with indefinite limit and the errors if any occurs for some period (say 1-3 months), then if we see any alarming degradation, we would have more data to be able to decide what can be a reasonable limit. we could also experiment there; e.g. change it down bit by bit and see what works best.
> 
> I agree. This would be easier if we have metrics on run durations of scheduled tasks. E.g. P50, P90, P99 of scheduled tasks durations. If we see a massive spike there then we know that we need to adjust this config setting. @megglos is that something we could ask from ZDP?

Isn't the batch size anyway limited by the record batch size limit? So we won't reach MAX_INT anyway as the record batch limit kicks in earlier, terminating the loop? So effectively only values that lead to a smaller record batch size than `maxAppendBatchSize = 32KB` will have an impact, everything beyond makes no difference as this limit kicks in earlier, di I got this right?

**Comment by oleschoenburg** [comment]
@megglos `maxAppendBatchSize` is not related here. I think you mean `maxMessageSize` because that's what limits how much the engine can write at once.

**Comment by korthout** [comment]
@oleschoenburg @megglos Yes there is a cap due to max message size, but that's not relevant. MAX_INT was chosen to not break behavior for message TTL checker while allowing to tweak the config option later. Lowering the value to some arbitrary number (like 1000) would change how long the scheduled tasks actor is blocked by a single pass of that checker. However, we have no way to easily detect (or measure) how long the scheduled tasks actually run and so we cannot easily select a fitting number. This PR is in the same situation as the message ttl checker's batch limit.

I'm therefore suggesting a new metric `scheduled-task-duration` to track how long scheduled tasks run and whether any of them block the actor excessively.

**Comment by megglos** [comment]
> @oleschoenburg @megglos Yes there is a cap due to max message size, but that's not relevant. MAX_INT was chosen to not break behavior for message TTL checker while allowing to tweak the config option later. Lowering the value to some arbitrary number (like 1000) would change how long the scheduled tasks actor is blocked by a single pass of that checker. However, we have no way to easily detect (or measure) how long the scheduled tasks actually run and so we cannot easily select a fitting number. This PR is in the same situation as the message ttl checker's batch limit.
> 
> I'm therefore suggesting a new metric `scheduled-task-duration` to track how long scheduled tasks run and whether any of them block the actor excessively.

Agree, would be in favor of assessing that in a follow-up then. @mustafadagher can you take care of creating that follow-up issue?

**Review by korthout — APPROVED** [review]
Thanks @mustafadagher for this great fix! 💯 

🎉 We tested in manually one last time and saw that the problem is resolved! We can now activate jobs with a timeout of 1s and activate them again a second later!

LGTM 👍 

Congratulations on your first major contribution to Zeebe! 👏

**Comment by mustafadagher** [comment]
> > @oleschoenburg @megglos Yes there is a cap due to max message size, but that's not relevant. MAX_INT was chosen to not break behavior for message TTL checker while allowing to tweak the config option later. Lowering the value to some arbitrary number (like 1000) would change how long the scheduled tasks actor is blocked by a single pass of that checker. However, we have no way to easily detect (or measure) how long the scheduled tasks actually run and so we cannot easily select a fitting number. This PR is in the same situation as the message ttl checker's batch limit.
> > I'm therefore suggesting a new metric `scheduled-task-duration` to track how long scheduled tasks run and whether any of them block the actor excessively.
> 
> Agree, would be in favor of assessing that in a follow-up then. @mustafadagher can you take care of creating that follow-up issue?

Done here https://github.com/camunda/zeebe/issues/17002

Let me know if I missed any details in the issue description.

**Comment by korthout** [comment]
@mustafadagher In the case where there are conflicts on all backport attempts, it's a good trick to try to manually backport to the latest version only (8.4) and then use backport-action again to attempt another auto-backport to the other branches.

**Pull request merged**
